tls/s2n_crl.c

```diff
@@ -86,3 +86,139 @@ int s2n_crl_get_issuer_hash(struct s2n_crl *crl, uint64_t *hash) {
+    for (uint32_t i = 0; i < num_contexts; i++) {
+        struct s2n_crl_lookup_context *context = NULL;
+        RESULT_GUARD(s2n_array_get(validator->crl_lookup_contexts, i, (void **) &context));
+        RESULT_ENSURE_REF(context);
+
+        if (!context->crl) {
+            /* A CRL was intentionally not returned from the callback. Don't add anything to the store.*/
+            continue;
+        }
+
+        RESULT_ENSURE_REF(context->crl->crl);
+        if (!sk_X509_CRL_push(validator->crl_stack, context->crl->crl)) {
+            RESULT_BAIL(S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+        }
+    }
+
+    X509_STORE_CTX_set0_crls(validator->store_ctx, validator->crl_stack);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_crl_get_lookup_callback_status(struct s2n_x509_validator *validator, crl_lookup_callback_status *status) {
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(validator->crl_lookup_contexts);
+
+    *status = FINISHED;
+
+    uint32_t num_contexts = 0;
+    RESULT_GUARD(s2n_array_num_elements(validator->crl_lookup_contexts, &num_contexts));
+    for (uint32_t i = 0; i < num_contexts; i++) {
+        struct s2n_crl_lookup_context *context = NULL;
+        RESULT_GUARD(s2n_array_get(validator->crl_lookup_contexts, i, (void **) &context));
+        RESULT_ENSURE_REF(context);
+
+        switch (context->status) {
+            case FINISHED:
+                break;
+            case AWAITING_RESPONSE:
+                *status = AWAITING_RESPONSE;
+                return S2N_RESULT_OK;
+        }
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_crl_handle_lookup_callback_result(struct s2n_x509_validator *validator) {
+    RESULT_ENSURE_REF(validator);
+
+    crl_lookup_callback_status status = 0;
+    RESULT_GUARD(s2n_crl_get_lookup_callback_status(validator, &status));
+    switch (status) {
+        case FINISHED:
+            RESULT_GUARD(s2n_crl_load_crls_from_contexts(validator));
+            validator->state = READY_TO_VERIFY;
+            break;
+        case AWAITING_RESPONSE:
+            validator->state = AWAITING_CRL_CALLBACK;
+            RESULT_BAIL(S2N_ERR_ASYNC_BLOCKED);
+        default:
+            RESULT_BAIL(S2N_ERR_INVALID_CERT_STATE);
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_crl_invoke_lookup_callbacks(struct s2n_connection *conn, struct s2n_x509_validator *validator) {
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(validator->cert_chain_from_wire);
+
+    DEFER_CLEANUP(struct s2n_array *crl_lookup_contexts = s2n_array_new(sizeof(struct s2n_crl_lookup_context)),
+                  s2n_array_free_p);
+    RESULT_ENSURE_REF(crl_lookup_contexts);
+
+    int cert_count = sk_X509_num(validator->cert_chain_from_wire);
+
+    for (int i = 0; i < cert_count; ++i) {
+        struct s2n_crl_lookup_context* context = NULL;
+        RESULT_GUARD(s2n_array_pushback(crl_lookup_contexts, (void**) &context));
+
```

Comment:
This doesn't seem as efficient as it could be. Arrays currently start with 16 entries, and have to alloc again if they resize. We don't need any resize behavior and don't need to guess at the size, because we already know the exact size of our array. I see two options to make this more efficient:

#### Use a C pointer / array

You could just allocate a `struct s2n_crl_lookup_context*` of the size you need. Unfortunately, you'd need to manually track its length, and the alloc/free would have to go through a blob due to our rules.

So you'd probably create a structure like:
```
struct s2n_crl_lookup_contexts {
    struct s2n_crl_lookup_context *contexts;
    const size_t count;
};
```
You can either recreate the s2n_blob when you need to free the array, or store it on the struct. I'd probably just recreate it.

#### Modify s2n_array to allow a set size
If you want to use s2n_array, than you'll need to make sure it's a) not constantly allocing new memory as it resizes b) not initially allocating more memory than you need. There's currently no way to do that, because that's not really the use case for the data structure. But you could add a new init/new method to start the array at a certain size, and either trust that if you choose the right size it will never resize or add a mechanism to enforce that.

Reply:
I decided to try modifying s2n_array: https://github.com/aws/s2n-tls/pull/3554

If this PR makes sense, I will update this PR to set the array capacity properly and set the array as static.

Reply:
I'm now using `s2n_array_new_with_capacity` here: c25d6a8b8ba055fc0624a88bbaa1c729343f5df0